### PR TITLE
Fix issues with `WagtailAdminFormPageForm.clean`

### PR DESCRIPTION
**Pull request by lb-**
* Ensure `WagtailAdminFormPageForm.clean` is applied regardless of related name.
* Fix & document support for extending `WagtailAdminFormPageForm` for form page admin validation.

Rebase & added unit tests from #10375 - includes clean up of docs with refined wording and a change that has a more streamlined code diff.

## Details

- Dynamically detect the `related_name` of any form fields attached to the page so that validation is actually run for custom related_names.
- This also will apply validation even if there are multiple `AbstractFormField` subclasses related to the page.
- Return `cleaned_data` for more consistent subclassing.
- Updated documentation with an example of adding custom page validation for form fields.
- Add unit test to ensure that the documented usage of extending `WagtailAdminFormPageForm` works as expected.
- Add unit test to ensure that duplicate clean name validation correctly runs on FormPage models with a custom related name.

### Recommended release notes

 * --FIXES--
 * Ensure form builder supports custom admin form validation (John-Scott Atlakson, LB (Ben) Johnston)
 * Ensure form builder correctly checks for duplicate field names when using a custom related name (John-Scott Atlakson, LB (Ben) Johnston)
 * --DOCS--
 * Document usage of [Custom validation for admin form pages](form_builder_custom_admin_validation) when using `AbstractEmailForm` or `AbstractForm` pages (John-Scott Atlakson, LB (Ben) Johnston)
 * --UPGRADE CONSIDIERATIONS--
 * For any pages that exist (e.g. `AbstractEmailForm`, `AbstractForm`, `FormMixin`, `EmailFormMixin`) that have a form field related name of anything other than `related_fields`. If these pages have duplicate field names, these will now be flagged as an error on saving after upgrading. Previously, these duplicate fields would have not shown in the form page, but should be considered.

### A. Custom related name duplicate field validation

1. Create a new form page model that uses a custom related name, not just `form_fields` as the default documentation recommends.
3. Attempt to create a new page with that model that has two fields with the same name (e.g. `subject`), observe that you do NOT get an error as expected.

Observe page can be created with duplicate field names (subject).

Screenshot 2024-10-28 at 6 39 33 PM

Instead we should get this error
Screenshot 2024-10-28 at 6 42 32 PM

```py

class SecondaryFormField(AbstractFormField):
    page = ParentalKey("SecondFormPage", related_name="secondary_form_fields", on_delete=models.CASCADE)


class SecondFormPage(AbstractEmailForm):

    content_panels = AbstractEmailForm.content_panels + [
        InlinePanel("secondary_form_fields", heading="Form fields", label="Field"),
        MultiFieldPanel(
            [
                FieldRowPanel(
                    [
                        FieldPanel("from_address"),
                        FieldPanel("to_address"),
                    ]
                ),
                FieldPanel("subject"),
            ],
            "Email",
        ),
    ]
```

### B. Custom admin form validation

1. In bakerydemo, add custom validation to the existing contact us model (code below)
2. Check that you can correctly do some basic validation or at least print out the cleaned_data in the console.

Current error encountered

Screenshot 2024-10-28 at 6 40 39 PM

Instead we should be able to read the `cleaned_data`.

```
from wagtail.models import Page
from wagtail.contrib.forms.models import FormMixin, WagtailAdminFormPageForm


class CustomWagtailAdminFormPageForm(WagtailAdminFormPageForm):
    def clean(self):
        cleaned_data = super().clean()
        # print the output or attempt to read this, you will get an error on `main` but should get the form submission data with this PR
        print(cleaned_data["from_address"])
        return cleaned_data


class FormPage(AbstractEmailForm):
    image = models.ForeignKey(
        "wagtailimages.Image",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="+",
    )
    body = StreamField(BaseStreamBlock(), use_json_field=True)
    thank_you_text = RichTextField(blank=True)

    base_form_class = CustomWagtailAdminFormPageForm ## add this line
    #...
```

### CODE REVIEW

**Comment by lb-** [comment]
I assumed I could push to the other branch, hence the inclusion of changelogs. 

Anyway, it would be great to get a review soon if possible @gasman / @thibaudcolas 

Otherwise I'm happy to wait and remove the changelogs.

File: wagtail/test/testapp/models.py [hunk 1]
```diff
@@ -715,18 +715,37 @@ def render_landing_page(self, request, form_submission=None, *args, **kwargs):
...
-        * Have custom related_name (see `FormFieldWithCustomSubmission.page`)
-        * Saves reference to a user
-        * Doesn't render html form, if submission for current user is present
+    A ``FormPage`` with a custom FormSubmission and other extensive customizations:
+
+    * A custom submission model
+    * A custom related_name (see `FormFieldWithCustomSubmission.page`)
+    * Saves reference to a user
+    * Doesn't render html form, if submission for current user is present
+    * A custom clean method that does not allow the 'from email' to be set to anything including example.com
```
Comment (lb-) [inline wagtail/test/testapp/models.py:741]:
```suggestion
    * A custom clean method that does not allow the 'from_address' to be set to anything including example.com
```

**Comment by lb-** [comment]
Just rebased, split the commit into two (for easier reviewing & cleaner isolated fixes) and removed changelog.

Will await a review.

**Comment by lb-** [comment]
Note that, all current Wagtail setups that do have a different related field name and that have saved duplicate fields will behave as follows.

1. The page will let them save duplicate fields
2. The form shown in the view (public facing page) will only show one field
3. When saved, the value from the single field will be applied to BOTH values and appear twice in the reports
4. Once this PR is merged, future applications will NOT be allowed to save both fields and will be forced to delete one of them.
5. No data will be lost, as the data would have been duplicated across the same named field anyway.

> Observe, two 'subject' fields but only one shown in the rendered form
Screenshot 2024-10-31 at 4 14 49 PM

Just wanted to better catalogue the bug as it stands today and what the 'upgrade considerations' need to be.

**Review by gasman — APPROVED** [review]
Looks good to me, thanks @lb- and @jsma!

**Comment by gasman** [comment]
Merged in 3013673c55a43fc25a7a480e5931a7d5d0a6fe23

**Comment by lb-** [comment]
Awesome, thanks Matt & thanks @jsma

**Comment by jsma** [comment]
Thanks @lb- for seeing this through to the finish line and @gasman for the review!
